src/block_service/directio_writable_file.cpp

```diff
@@ -0,0 +1,116 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+#include <cstring>
+#include <fcntl.h>
+#include <stdlib.h> // posix_memalign
+#include <string>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h> // getpagesize
+
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+
+#include "block_service/directio_writable_file.h"
+
+DSN_DEFINE_uint32("replication",
+                  direct_io_buffer_pages,
+                  1,
+                  "Number of pages we need to set to direct io buffer");
+DSN_TAG_VARIABLE(direct_io_buffer_pages, FT_MUTABLE);
+
+DSN_DEFINE_bool("replication",
+                enable_direct_io,
+                false,
+                "Whether to enable direct I/O when download files");
+DSN_TAG_VARIABLE(enable_direct_io, FT_MUTABLE);
+
+namespace dsn {
+namespace dist {
+namespace block_service {
+
+const uint32_t g_page_size = getpagesize();
+
+direct_io_writable_file::direct_io_writable_file(const std::string &file_path)
+    : _fd(-1),
+      _file_size(0),
+      _buffer(),
+      _buffer_size(FLAGS_direct_io_buffer_pages * g_page_size),
+      _offset(0)
+{
+    if (posix_memalign(&_buffer, g_page_size, _buffer_size) != 0) {
+        derror_f("Allocate memaligned buffer failed, errno = {}", errno);
```

Comment:
How will the server act when the ctor failed?

Reply:
OK, I add a init function.